code/ATMOSPHERICS/atmospherics.dm

```diff
@@ -277,35 +279,54 @@ Pipelines + Other Objects -> Pipe network
 
 	var/obj/machinery/atmospherics/target_move = findConnecting(direction)
 	if(target_move)
-		if(is_type_in_list(target_move, GLOB.ventcrawl_machinery) && target_move.can_crawl_through())
-			user.remove_ventcrawl()
-			user.forceMove(target_move.loc) //handles entering and so on
-			user.visible_message("You hear something squeezing through the ducts.", "You climb out the ventilation system.")
+		if(is_type_in_list(target_move, GLOB.ventcrawl_machinery))
+			user.try_crawl_out(target_move)
 		else if(target_move.can_crawl_through())
-			if(returnPipenet() != target_move.returnPipenet())
-				user.update_pipe_vision(target_move)
 			user.loc = target_move
 			user.client.eye = target_move //if we don't do this, Byond only updates the eye every tick - required for smooth movement
+			if(returnPipenet() != target_move.returnPipenet()) // This forces a redraw of the pipes for the ventcrawler
+				user.update_pipe_vision()
 			if(world.time - user.last_played_vent > VENT_SOUND_DELAY)
 				user.last_played_vent = world.time
 				playsound(src, 'sound/machines/ventcrawl.ogg', 50, 1, -3)
 	else
 		if((direction & initialize_directions) || is_type_in_list(src, GLOB.ventcrawl_machinery)) //if we move in a way the pipe can connect, but doesn't - or we're in a vent
-			user.remove_ventcrawl()
-			user.forceMove(src.loc)
-			user.visible_message("You hear something squeezing through the pipes.", "You climb out the ventilation system.")
-	user.canmove = 0
+			if(do_after(user, 30, target = src))
+				user.remove_ventcrawl()
+				user.forceMove(src.loc)
+				user.visible_message("<span class='boldnotice'>You hear something squeezing through the pipes.</span>", "<span class='notice'>You climb out the ventilation system.</span>")
+			else
+				to_chat(user, "<span class='notice'>You stop climbing out of the ventilation system.</span>")
+	user.canmove = FALSE
 	spawn(1)
-		user.canmove = 1
+		user.canmove = TRUE
 
 /obj/machinery/atmospherics/AltClick(var/mob/living/L)
-	if(is_type_in_list(src, GLOB.ventcrawl_machinery))
+	if(is_type_in_list(src, GLOB.ventcrawl_machinery) || check_open())
 		L.handle_ventcrawl(src)
 		return
 	..()
 
 /obj/machinery/atmospherics/proc/can_crawl_through()
-	return 1
+	return TRUE
+
+/obj/machinery/atmospherics/proc/check_open()
+	for(var/direction in GLOB.cardinal)
+		if((direction & initialize_directions) && !findConnecting(direction)) // Checks if a pipe can connect in a direction, then returns true if there isn't a pipe connected there
+			return TRUE
+	return FALSE
+
+/obj/machinery/atmospherics/proc/update_all_pipe_vision()
+	for(var/mob/living/M in GLOB.ventcrawlers)
```

Comment:
Please try to make this typeless

Reply:
Vent crawl code is defined on a mob living level, so I'm not quite sure what you mean here.